x-pack/plugins/reporting/server/config/index.ts

```diff
@@ -0,0 +1,88 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import { i18n } from '@kbn/i18n/';
+import { TypeOf } from '@kbn/config-schema';
+import crypto from 'crypto';
+import { map } from 'rxjs/operators';
+import { PluginConfigDescriptor } from 'kibana/server';
+import { CoreSetup, Logger, PluginInitializerContext } from '../../../../../src/core/server';
+import { ConfigSchema, ConfigType } from './schema';
+
+/*
+ * Set up dynamic config defaults
+ */
+export function createConfig$(core: CoreSetup, context: PluginInitializerContext, logger: Logger) {
+  return context.config.create<TypeOf<typeof ConfigSchema>>().pipe(
+    map(config => {
+      // encryption key
+      let encryptionKey = config.encryptionKey;
+      if (encryptionKey === undefined) {
+        logger.warn(
+          i18n.translate('xpack.reporting.serverConfig.randomEncryptionKey', {
+            defaultMessage:
+              'Generating a random key for xpack.reporting.encryptionKey. To prevent sessions from being invalidated on ' +
+              'restart, please set xpack.reporting.encryptionKey in kibana.yml',
+          })
+        );
+        encryptionKey = crypto.randomBytes(16).toString('hex');
+      }
+
+      const { kibanaServer: reportingServer } = config;
+      const serverInfo = core.http.getServerInfo();
+
+      // kibanaServer.hostname, default to server.host, don't allow "0"
+      let kibanaServerHostname = reportingServer.hostname
+        ? reportingServer.hostname
+        : serverInfo.host;
+      if (kibanaServerHostname === '0') {
+        logger.warn(
+          i18n.translate('xpack.reporting.serverConfig.invalidServerHostname', {
+            defaultMessage:
+              `Found 'server.host: "0" in Kibana configuration. This is incompatible with Reporting. ` +
+              `To enable Reporting to work, '{configKey}: 0.0.0.0' is being automatically to the configuration. ` +
+              `You can change the setting to 'server.host: 0.0.0.0' or add '{configKey}: 0.0.0.0' in kibana.yml to prevent this message.`,
+            values: { configKey: 'xpack.reporting.kibanaServer.hostname' },
+          })
+        );
+        kibanaServerHostname = '0.0.0.0';
+      }
+
+      // kibanaServer.port, default to server.port
+      const kibanaServerPort = reportingServer.port
+        ? reportingServer.port
+        : serverInfo.port; // prettier-ignore
+
+      // kibanaServer.protocol, default to server.protocol
+      const kibanaServerProtocol = reportingServer.protocol
+        ? reportingServer.protocol
+        : serverInfo.protocol;
+
+      return {
```

Comment:
@joelgriffith I need to hold off on this PR for a little bit. I just realized one thing this piece is missing is the replacement for setting a default for the `xpack.reporting.capture.browser.chromium.disableSandbox` setting. 

https://www.elastic.co/guide/en/kibana/current/reporting-settings-kb.html#reporting-chromium-settings
> Defaults to false for all operating systems except Debian, Red Hat Linux, and CentOS which use true

I think that takes care of all the parts we had on config defaults set at run-time.